### PR TITLE
Ensemble stacking integration in chap_core

### PR DESCRIPTION
**Pull request by behdadnikkhah25**
# Fix deterministic stacking, NNLS weighting, and sample collapsing

## Summary

This PR finishes the ensemble redesign and addresses the main review comments by:

- making the **deterministic ensemble a clean point-only stacking model** (no residual bootstrap),
- fixing **NNLS usage** so raw coefficients are used for prediction and only normalized for reporting,
- making **probabilistic → deterministic collapsing** explicit (median + warnings),
- documenting and testing the **vincentization** scheme for probabilistic ensembles, and
- improving logging and tests around data issues and fallbacks.

It also:

- removes legacy wrappers in favor of `wrappers.py`,
- moves ensemble config YAMLs to `tests/fixtures/ensemble_config/`,
- removes the unused `use_residual_bootstrap` flag from models and CLI, and
- keeps all tests passing (`964 passed, 117 skipped, 4 xfailed, 1 xpassed`).

---

### Deterministic ensemble = clean stacking (no bootstrap)

- Residual bootstrap has been **entirely removed** from:
  - `EnsembleModel` / `EnsembleEstimator`, and
  - the ensemble CLI.
- Deterministic ensembles now:
  - output a **single sample** per forecast (`sample_0`),
  - log a warning that **CRPS reduces to MAE** and is not directly comparable to multi-sample CRPS.
- `use_residual_bootstrap` is **removed** from:
  - `EnsembleModel`,
  - `EnsembleEstimator`, and
  - CLI signatures and tests.

### NNLS coefficients: prediction vs reporting

- `NonNegativeMetaModel.fit` stores the raw NNLS solution in `coef_`.
- `NonNegativeMetaModel.predict` uses `X @ coef_` directly (no renormalization).
- `EnsembleModel.train`:
  - reads `coef_`, clips to non-negative,
  - normalizes **only for logging/reporting** as percentages in `self.weights`.
- This avoids the previous “optimize one objective, predict with another” behavior.

### Probabilistic → deterministic collapse (median + warnings)

- When probabilistic base models are used in the deterministic ensemble:
  - `SampleExtractor.samples_to_flat` detects `sample_*` columns,
  - logs a warning that **uncertainty is discarded**,
  - collapses to a point forecast via the **median** across samples per row.
- This replaces the previous mean-based, silent collapse.

### Probabilistic meta-model: vincentization

- `ProbabilisticMetaModel` now combines base distributions via **vincentization (quantile averaging)**:
  - per row, each base’s samples approximate $Q_i(p)$,
  - samples are sorted and combined as

    ```math
    Q_{\text{ens}}(p) = \sum_i w_i Q_i(p),
    ```

    with non-negative simplex weights $w_i$.

- Implementation:
  - `_vincentize_samples`:
    - validates shapes,
    - stacks samples as `(n_models, n_rows, n_samples)`,
    - sorts along the sample axis and applies weights.
  - `fit`:
    - optimizes weights on the simplex with SLSQP using `crps_score_unbiased_matrix` as objective,
    - falls back to uniform weights with a warning if optimization fails.
  - `predict`:
    - applies `_vincentize_samples`,
    - clips negative ensemble values at zero.

- Resulting samples are **sorted quantiles** per row (monotone), not arbitrary Monte Carlo draws, and CRPS is evaluated with the same unbiased helper used by the metrics.

### Observability: NaNs, fallbacks, CRPS semantics

- During meta-model training:
  - rows with NaNs in targets or features are dropped,
  - total dropped rows and **per-base NaN counts** are logged.
- If NNLS yields non-positive weights:
  - a warning is logged,
  - `self.weights` is set to uniform percentages.
- When deterministic ensembles output a single sample:
  - a warning explicitly states that **CRPS = MAE**.
- When probabilistic samples are collapsed:
  - a warning states that **probabilistic information is discarded**.

---

### `chap_core/ensemble/_meta_models.py`

- `NonNegativeMetaModel`:
  - raw NNLS `coef_` used for prediction.
- `ProbabilisticMetaModel`:
  - vincentization via `_vincentize_samples`,
  - SLSQP optimization on the simplex with `crps_score_unbiased_matrix`,
  - uniform fallback on failure.
- `crps_ensemble`:
  - delegates to `crps_score_unbiased_matrix`.

### `chap_core/ensemble/ensemble_model.py`

- Inner split uses `train_test_split` on `TimePeriod`.
- Probabilistic mode:
  - uses `SampleExtractor.reshape_samples` and `ProbabilisticMetaModel`.
- Deterministic mode:
  - uses `SampleExtractor.samples_to_flat` and `NonNegativeMetaModel`.
- Logs:
  - NaN drops and per-base NaNs,
  - normalized meta-weights as percentages.
- `use_residual_bootstrap` removed.

### `chap_core/ensemble/_predictor.py`

- Probabilistic:
  - aligns samples via `reshape_samples`, calls `meta.predict`, repacks with `_pack_samples`.
- Deterministic:
  - builds meta-features, calls `NonNegativeMetaModel.predict`,
  - logs CRPS=MAE warning,
  - returns single-sample `Samples` (`sample_0`).

### `chap_core/ensemble/_sample_extractor.py`

- `samples_to_flat`:
  - uses `forecast`/`value` if present,
  - otherwise collapses `sample_*` with **median** and logs a warning.
- `reshape_samples`:
  - aligns on `(location, time_period)`,
  - tiles point forecasts when samples are missing, with warning,
  - resamples or tiles to `target_n` samples as needed.

### `chap_core/ensemble/wrappers.py`

- Replaces legacy wrappers with:
  - `BaseModelSpec(template, config)`,
  - `TemplateWithConfig` that forwards `config` to `template.get_model(config)` and exposes `name`/`repo`.

### `chap_core/cli_endpoints/ensemble.py`

- CLI for `evaluate-ensemble`:
  - builds `TemplateWithConfig` list from `--base-model-names` and optional `--model-configuration-yaml`,
  - creates an `EnsembleModel`,
  - runs `Evaluation.create` and writes NetCDF + CSV reports,
  - writes `ensemble_meta_report.csv` with base model weights (percentages).
- `use_residual_bootstrap` removed from the CLI API.

### `chap_core/cli_endpoints/_common.py`

- Shared CLI helpers, including a **temporary** `save_results` patch to support the new evaluation outputs.
- Explicitly marked as temporary and to be removed once upstream provides a unified solution.

---

## Tests

All tests pass:

- Lint & type checking:

```bash
  uv run make lint
  # Ruff: all checks passed
  # Mypy: Success: no issues found in 382 source files
  
```

- Test suite:

```bash
uv run make test
# 964 passed, 117 skipped, 4 xfailed, 1 xpassed, 7 warnings

```

Key updated/added areas:

- tests/ensemble/test_ensemble_model.py:
inner split, input validation, NaN logging.
- tests/ensemble/test_ensemble_stacking.py:
deterministic stacking (single sample, weights reporting) and probabilistic sample shapes/monotonicity.
- tests/ensemble/test_predictor.py:
deterministic CRPS=MAE warning, probabilistic sample shapes.
- tests/ensemble/test_meta_models.py:
NNLS behavior, simplex constraints, vincentization invariance, fallback.
- tests/ensemble/test_sample_extractor.py:
median collapse and alignment/resampling in reshape_samples.
- tests/ensemble/test_legacy_wrappers.py:
BaseModelSpec and TemplateWithConfig.
- tests/cli_endpoints/test_ensemble_cli.py:
CLI wiring for deterministic/probabilistic ensemble evaluation.
- tests/fixtures/ensemble_config/*.yaml
ensemble configs moved into fixtures.

#### 1) Deterministic ensemble (point-only, no bootstrap)
```bash
chap evaluate-ensemble \
  --base-model-names https://github.com/chap-models/rwanda_sarimax,https://github.com/chap-models/chap_auto_ewars,https://github.com/chap-models/INLA_baseline_model \
  --model-configuration-yaml ../tests/fixtures/ensemble_config/rwanda_sarimax_config.yaml,../tests/fixtures/ensemble_config/chap_ewars_monthly_config.yaml,../tests/fixtures/ensemble_config/inla_baseline_config.yaml \
  --ensemble-method deterministic \
  --random-state 20 \
  --dataset-csv https://raw.githubusercontent.com/dhis2/climate-health-data/main/lao/chap_LAO_admin1_monthly.csv \
  --report-filename ensemble_report.csv \       
  --output-file ensemble_report.nc \       
  --backtest-params.n-splits 3 \
  --backtest-params.n-periods 3

```

#### 2) Probabilistic ensemble
```bash
chap evaluate-ensemble \                                                                              
  --base-model-names https://github.com/chap-models/rwanda_sarimax,https://github.com/chap-models/chap_auto_ewars,https://github.com/chap-models/INLA_baseline_model \
  --model-configuration-yaml ../tests/fixtures/ensemble_config/rwanda_sarimax_config.yaml,../tests/fixtures/ensemble_config/chap_ewars_monthly_config.yaml,../tests/fixtures/ensemble_config/inla_baseline_config.yaml \
  --ensemble-method probabilistic \
  --random-state 20 \
  --dataset-csv https://raw.githubusercontent.com/dhis2/climate-health-data/main/lao/chap_LAO_admin1_monthly.csv \
  --report-filename ensemble_report.csv \
  --output-file ensemble.nc \
  --backtest-params.n-splits 3 \
  --backtest-params.n-periods 3

```

### Out of scope / deferred
- Aligning evaluate-ensemble CLI UX fully with other commands (future PR).
- Removing the temporary save_results patch in _common.py once upstream has a shared solution for report writing.

### CODE REVIEW

**Comment by knutdrand** [comment]
Superseded by #534.

That PR carries this branch's history and commits unchanged, merged onto current master with the review findings fixed on top. Opened against upstream so CI runs — this PR is from a fork's `master`, so the workflows never triggered here.

Worth knowing: merging this branch as-is onto current master fails 19 tests in `tests/evaluation/test_metrics.py`. Master vectorised CRPS in #467 after this branch was cut, and this branch also switches the estimator from the biased `1/m^2` form to the fair `1/(m(m-1))` form, which the lock-in tests catch. #534 keeps master's definition and adds a public `crps_matrix` helper for the ensemble's batch scoring.

The other substantive fix is target leakage: `disease_cases` was reaching base models during inner validation via the future-data CSV, so the meta-weights were fitted against values the base models could see.

Full details in #534.

**Comment by knutdrand** [comment]
Hi — first, an apology: this PR sat for three months without review, and that's on us, not you. You'd already addressed the earlier round of feedback on #325 and then got nothing back.

I've opened #534, which is your branch and your commits merged onto current master with some fixes on top. I did it that way so CI would actually run — this PR is from your fork's `master`, so the workflows never triggered, which is why nothing caught that it now fails 19 tests in `tests/evaluation/test_metrics.py`.

Two real bugs I found and fixed:

- `disease_cases` was reaching the base models through the future-data CSV during inner validation, so the meta-weights were being fitted against values the models could see.
- The CRPS rewrite switched the estimator to the fair `1/(m(m-1))` form, which master's tests lock against the biased `1/m^2` form. Master also vectorised CRPS itself in #467 after you branched, so that part of the work is now redundant.

One change is a judgment call I made on your behalf and would like your view on: I changed inner validation to rolling windows of `n_periods`, because the weights were being fitted at a 12-step horizon and then applied at 3. If you had a reason for the single long window, say so and I'll revert it.

Your design is unchanged otherwise — vincentization, the NNLS/SLSQP split and the CLI shape are all as you had them.

Could you review #534? You know this code better than I do.

**Comment by behdadnikkhah25** [comment]
I know you’ve had a lot on your plate and haven’t had much time to look at this – I really appreciate you taking the time now (and also fixing important bugs along the way). I hope you’re available to look at it a bit now so we can land this. As I’ve mentioned, I’m a student and I’m stressing a bit to get this in so I can focus on testing, evaluating, and writing the thesis. I’m hoping we can land this together in chap_core so both of us get value from it.

I’ve pulled the branch locally and run evaluate-ensemble end-to-end on the LAO admin1 monthly dataset with three external base models (rwanda_sarimax, chap_auto_ewars, INLA_baseline), using the script I shared. Both probabilistic and deterministic modes exit 0. I also ran each base model individually with the same backtest config and compared everything via chap export-metrics.

---

## Probabilistic ensemble (LAO admin1 monthly, 3 splits, horizon 3)

Inner validation: 151 periods, rolling-origin windows at n_periods=3, n_splits=3, stride=1, dropping rows with NaNs in targets/features.

Meta-weights:
- rwanda_sarimax: ≈11.5%
- chap_ewars_monthly: ≈2.9%
- INLA_baseline: ≈85.6%

Metrics:
- CRPS ≈ 83.14, MAE ≈ 117.30, RMSE ≈ 247.31
- Coverage 10–90 ≈ 0.71, Coverage 25–75 ≈ 0.49
- sample_count = 14 400 (100 samples per forecast)

CRPS is below MAE, and coverage/Winkler scores behave sensibly for a genuine predictive distribution.

---

## Deterministic ensemble

Base-model samples are collapsed to a single point forecast (median), as indicated by the “Collapsing probabilistic samples…” warnings.

Meta-weights:
- rwanda_sarimax: ≈75.5%
- chap_ewars_monthly: 0.0%
- INLA_baseline: ≈24.5%

Metrics:
- CRPS = MAE ≈ 177.62
- RMSE ≈ 291.15
- Coverage 10–90 = 0, Coverage 25–75 = 0
- sample_count = 144 (one sample per forecast)

The equality CRPS = MAE matches the “CRPS reduces to MAE” warning and confirms that the deterministic ensemble is behaving as a pure point-forecast meta-model.

---

## Base models (same dataset and backtest config)

From metrics_comparison_25_08.csv:

- rwanda_sarimax: CRPS ≈ 196.59, RMSE ≈ 324.27
- chap_auto_ewars: CRPS ≈ 91.94, RMSE ≈ 304.33
- INLA_baseline: CRPS ≈ 83.83, RMSE ≈ 269.19

These line up nicely with the ensemble weights: the probabilistic ensemble leans on INLA (best CRPS), while the deterministic ensemble leans on rwanda_sarimax (better point forecasts in this setup).

---

## Question about earlier evaluation

One question: when you evaluated the ensemble earlier (before my run with three base models), which two models did you use? It would be helpful to know exactly which ones and whether you saw similar weight/CRPS–MAE behaviour.

---

## Technical checks

On the technical side, I’ve checked that disease_cases is not present in the future-data CSVs used during inner validation, so the original target-leakage issue is fixed. Rejecting LA-VI due to missing targets over the training period also looks correct.

---

## Inner validation design

On the inner validation design: I still think your original idea of a single 12-period inner window is reasonable, but using rolling windows at the backtest horizon (n_periods = horizon) feels more consistent with how the ensemble is actually evaluated in CHAP. We validate the weights on the same kind of forecasts (same horizon and rolling-origin structure), and the logs (train=139, val=12, 4 windows of horizon 3) plus the resulting meta-weights reflect that.

---

## Review position / merge readiness

From my side, this PR is approved: the implementation matches the architecture/design I’ve been working towards, and the end-to-end evaluation on LAO with three base models looks solid. As far as my review is concerned, this is merge-ready into chap_core.

---

## Thesis workflow question

Also, for my thesis workflow: do you think I should already merge this branch into the master branch of my fork, and treat that as my “thesis baseline”, or would you prefer that I wait until this is merged upstream first?

**Comment by behdadnikkhah25** [comment]
The script for running the ensemble with the three chosen base_models: [run_ensemble_analysis.sh](https://github.com/user-attachments/files/31433016/run_ensemble_analysis.sh)
